rllib/agents/slateq/slateq_tf_policy.py

```diff
@@ -0,0 +1,356 @@
+    slate_sum_q_values = tf.reduce_sum(input_tensor=slate_q_values, axis=2)
+    return slate_sum_q_values
+
+
+def score_documents(
+    user_obs, doc_obs, no_click_score=1.0, multinomial_logits=False, min_normalizer=-1.0
+):
+    """Computes dot-product scores for user vs doc (plus no-click) feature vectors."""
+
+    # Dot product between used and each document feature vector.
+    scores_per_candidate = tf.reduce_sum(
+        tf.multiply(tf.expand_dims(user_obs, 1), tf.stack(doc_obs, axis=1)), 2
+    )
+    # Compile a constant no-click score tensor.
+    score_no_click = tf.fill([tf.shape(user_obs)[0], 1], no_click_score)
+    # Concatenate click and no-click scores.
+    all_scores = tf.concat([scores_per_candidate, score_no_click], axis=1)
+
+    # Logits: Softmax to yield probabilities.
+    if multinomial_logits:
+        all_scores = tf.nn.softmax(all_scores)
+    # Multinomial proportional model: Shift to `[0.0,..[`.
+    else:
+        all_scores = all_scores - min_normalizer
+
+    # Return click (per candidate document) and no-click scores.
+    return all_scores[:, :-1], all_scores[:, -1]
+
+
+def setup_early(policy, obs_space, action_space, config):
+    """Obtain all possible slates given current docs in the candidate set."""
+
+    num_candidates = action_space.nvec[0]
+    slate_size = len(action_space.nvec)
+
+    mesh_args = [list(range(num_candidates))] * slate_size
+    slates = tf.stack(tf.meshgrid(*mesh_args), axis=-1)
+    slates = tf.reshape(slates, shape=(-1, slate_size))
+    # Filter slates that include duplicates to ensure each document is picked
+    # at most once.
+    unique_mask = tf.map_fn(
+        lambda x: tf.equal(tf.size(input=x), tf.size(input=tf.unique(x)[0])),
+        slates,
+        dtype=tf.bool,
+    )
+    # slates.shape: [A, S]
+    slates = tf.boolean_mask(tensor=slates, mask=unique_mask)
+
+    # Store all possible slates only once in policy object.
+    policy.slates = slates
+
+
+def setup_mid_mixins(policy: Policy, obs_space, action_space, config) -> None:
+    """Call mixin classes' constructors before SlateQTorchPolicy loss initialization.
+
+    Args:
+        policy: The Policy object.
+        obs_space: The Policy's observation space.
+        action_space: The Policy's action space.
+        config: The Policy's config.
+    """
+    LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
+
+
+def setup_late_mixins(
+    policy: Policy,
+    obs_space: gym.spaces.Space,
+    action_space: gym.spaces.Space,
+    config: TrainerConfigDict,
+) -> None:
+    """Call mixin classes' constructors after SlateQTorchPolicy loss initialization.
+
+    Args:
+        policy: The Policy object.
+        obs_space: The Policy's observation space.
+        action_space: The Policy's action space.
+        config: The Policy's config.
+    """
+    TargetNetworkMixin.__init__(policy, config)
+
+
+def rmsprop_optimizer(
```

Comment:
we don't have an existing util file that defines all the optimizers?

Reply:
Nope :D

Reply:
We should make optimizers configurable as we do exploration/(soon) buffers/inputs:

```
config:
  optimizer_config:
    type: Adam|RMSprop|...
    lr: ...
    epsilon: ...
```